### ISSUE TITLE
ShaderCompiler.cpp(197,5): error C2440: 'initializing': cannot convert from 'initializer list' to 'int'

### ISSUE THREAD

**Issue by eslinan** [vsg-dev/VulkanSceneGraph#658]

**Comment by eslinan** [comment]
ShaderCompiler.cpp(197,5): error C2440: 'initializing': cannot convert from 'initializer list' to 'int'

**Comment by artemkulyk** [comment]
Please check https://github.com/vsg-dev/VulkanSceneGraph/issues/237
If you updated to a newer Vulkan SDK on macOS, try to run cmake again from scratch and rebuild. Possibly, it will help to fix the issue.

**Comment by robertosfield** [comment]
@eslinan what OS, Vulkan and GLSLang version are you compiling against?

**Comment by eslinan** [comment]
windows, vs2022, vulkan1.3.231.1

**Comment by eslinan** [comment]
glslang_INCLUDE_DIR = C:/VulkanSDK/1.3.231.1/Include
glslang_LIBRARY = C:/VulkanSDK/1.3.231.1/Lib/glslang.lib

**Comment by robertosfield** [comment]
This is pretty odd, in theory you should be compiling against the same GLSL headers to me, so perhaps something else is failing.  The next stone to turn over is see whether CMake setup is failing to detect the features in the GLSLang headers it should, if you look at the ShaderCompiler.cpp code preceding where you see the error:

https://github.com/vsg-dev/VulkanSceneGraph/blob/master/src/vsg/utils/ShaderCompiler.cpp#L193

You'll notice the code:

#    ifdef GLSLANG_ResourceLimits_maxDualSourceDrawBuffersEXT
    1, // maxDualSourceDrawBuffersEXT
#    endif

I'm wondering if GLSLANG_ResourceLimits_maxDualSourceDrawBuffersEXT isn't be set by CMake where it should be, so it's not compiling this line. This block exists as the project devs of GLSLang have kept changing this structure so different versions required different settings, but GLSLang doesn't provide any mechanism for know which is required.  So... to get the VSG to compile against all the GLSLang in the wild we have to use CMake to attempt to compile code snippets to detect different features.

The CMake script that does this is:
https://github.com/vsg-dev/VulkanSceneGraph/blob/master/src/vsg/CMakeLists.txt#L254

Perhaps this isn't being run, or failing when it should for some reason and GLSLANG_ResourceLimits_maxDualSourceDrawBuffersEXT isn't being set to 1.

In my CMakeCache.txt I have:

./CMakeCache.txt://Test GLSLANG_ResourceLimits_maxDualSourceDrawBuffersEXT
./CMakeCache.txt:GLSLANG_ResourceLimits_maxDualSourceDrawBuffersEXT:INTERNAL=1

Could you check your CMakeCache.txt to see what value it has.

**Comment by robertosfield** [comment]
@eslinan did you get any further?

I haven't seen reports from other Windows users about problems so it would be useful to get more information about your system.  What CMake version did you use?  

Could you post your CMakeCache.txt file?

**Comment by robertosfield** [comment]
I can't do anything more myself so I've created a [Discussion thread](668) to request assistance from Windows users to see if they can provide any further insight,.

**Comment by eslinan** [comment]
ok. my cmake version is 3.25.0.
[CMakeCache.txt](https://github.com/vsg-dev/VulkanSceneGraph/files/10250560/CMakeCache.txt)

**Comment by eslinan** [comment]
the error codes doesn't included between #if GLSLANG_ResourceLimits_maxDualSourceDrawBuffersEXT #endif
    {
        1, // nonInductiveForLoops
        1, // whileLoops
        1, // doWhileLoops
        1, // generalAttributeMatrixVectorIndexing
        1, // generalVaryingIndexing

        1, // generalVaryingIndexing
        1, // generalSamplerIndexing
        1, // generalVariableIndexing
        1  // generalConstantMatrixVectorIndexing
    }};

**Comment by eslinan** [comment]
the error codes doesn't included between #if GLSLANG_ResourceLimits_maxDualSourceDrawBuffersEXT #endif
    {
        1, // nonInductiveForLoops
        1, // whileLoops
        1, // doWhileLoops
        1, // generalAttributeMatrixVectorIndexing
        1, // generalVaryingIndexing

        1, // generalVaryingIndexing
        1, // generalSamplerIndexing
        1, // generalVariableIndexing
        1  // generalConstantMatrixVectorIndexing
    }};

**Comment by eslinan** [comment]
the error codes doesn't included between #if GLSLANG_ResourceLimits_maxDualSourceDrawBuffersEXT #endif
<img width="310" alt="1671243400563" src="https://user-images.githubusercontent.com/4784744/208218910-a0fe86dc-2d4f-4fbe-9d00-6cf795259506.png">
<img width="759" alt="1671243437715" src="https://user-images.githubusercontent.com/4784744/208218933-22efd5f8-7b58-4244-ac1b-62d5fdb54a9d.png">

**Comment by robertosfield** [comment]
Thanks for the CMakeCache.txt, I've compared the section relating to GLSLang tests to what I see on my Kubuntu 22.04 system with CMake 3.22.1 and VulkanSDK 1.3.231.1:

![Screenshot_20221217_182939](https://user-images.githubusercontent.com/1327952/208256283-c66d21b6-85a8-437c-8462-74b14f96d313.png)

To me it looks like the tests for GLSLang features is failing on @eslinan's CMake configuration.  I don't know why this is yet, and why other Windows users aren't seeing similar problems. Could be that recent CMake versions have changed how check_cxx_source_compiles functions?

The part of the VSG's CMake build system that tests GLSLang features is at:

https://github.com/vsg-dev/VulkanSceneGraph/blob/master/src/vsg/CMakeLists.txt#L262

Perhaps adding some extra debug messaging in the results of the check_cxx_source_compiles is required.  Or perhaps something extra to help the Windows detection side?

**Comment by 640kb** [comment]
Hello,

I have updated to CMake 3.25.0 myself and did not encounter the problem. I am using Windows, Visual Studio 2022 and the Vulkan SDK 1.3.231.1

Here my CMakeCache:

[CMakeCache.txt](https://github.com/vsg-dev/VulkanSceneGraph/files/10267151/CMakeCache.txt)

**Comment by robertosfield** [comment]
Thanks for the cache file @640kb.  I've done a xxdiff between the two files and the GLSLang test section is different between @eslinan's and yours.  

![Screenshot_20221220_094258](https://user-images.githubusercontent.com/1327952/208635384-0a2c76e2-75e6-4cb5-abe6-7d60d8222423.png)

This section of yours is closer to what I see under Linux.

One difference I do see that might be important is the different Makefiles generated, NMake vs Visual Studio i.e.

```
//Name of generator.
CMAKE_GENERATOR:INTERNAL=Visual Studio 17 2022

```
vs

```
//Name of generator.
CMAKE_GENERATOR:INTERNAL=NMake Makefiles JOM

```

Could it be the compiler or the makefile used is the root of the problem?

**Comment by 640kb** [comment]
I just tried this CMake generator. I had some problems (but I don't know the Visual Studio IDE very well), but after a few tries I was able to compile VulkanSceneGraph without errors. So the vsgd.lib could be created.

CMakeCache:
[CMakeCache.txt](https://github.com/vsg-dev/VulkanSceneGraph/files/10267425/CMakeCache.txt)

But I think the generated Visual Studio project files are not working quite correctly. When I want to compile the complete project, it doesn't work. You have to be careful in the IDE to create only VulkanSceneGraph. Maybe someone else can say something about it.

**Comment by sbrkopac** [comment]
I just rebuilt my VSG ecosystem from scratch on Windows, and all build targets worked out of the box. [CMakeCache.txt](https://github.com/vsg-dev/VulkanSceneGraph/files/10282940/CMakeCache.txt)

**Comment by robertosfield** [comment]
Thanks @sbrkopac, I've done a diff against the CMakeCache for the failed build and again the difference is the extra 
![Screenshot_20221222_174612](https://user-images.githubusercontent.com/1327952/209195179-ddb8481c-f0e7-42d1-b013-29aabb50fc45.png) // Result of TRY_COMPILE in the failed build case.

Both builds use the same VS version, but different CMake versions.

However, the other successful Windows build uses the same CMake version, but differs in using NMake build.

I'm stumped as why one CMake run has problems but others don't.  Could it be how Cmake is invoked?  @eslinan, @640kb and @sbrkopac  how did you run CMake to generate CMakeCache.txt and associated build files?

**Comment by sbrkopac** [comment]
For Visual Studio project files - I use cmake-gui and it just works out of the box. I only update the install path.

I tested NMake Makefiles: [CMakeCache.txt](https://github.com/vsg-dev/VulkanSceneGraph/files/10289942/CMakeCache.txt). The command I run was ```cmake .. -G "NMake Makefiles"```. Seems to work out of the box:

![image](https://user-images.githubusercontent.com/1939712/209215561-28fe43da-a2f2-46db-806c-335aca8656e1.png)

**Comment by robertosfield** [comment]
Happy New Year folks :-)

I've reviewed the VulkanSceneGraph/src/vsg/CMakeLists.txt again and spotted the check_cxx_source_compiles() was using ${glslang_LIBRARY} rather than the more inclusive ${glslang_LIBRARIES} so I've amended  it to use the later:

d51a369137aafc09c9918f489fd5dc623c9881c6

I don't see any differences in my CMakeCache.txt but then it's been working all the way along.  I haven't yet found a reason why @eslinan build has TRY_COMPILE and TRY_RUN entries while no one else appears to have this.  Could this CMake being configured differently?

The above chance is spin of the wheel, I really don't know if it might make any difference but it seems like a sensible enough change so perhaps @eslinan can try VSG master and see if it affects things.

**Comment by robertosfield** [comment]
I have spent the last week working on resolving the issues with glslang/SPRIV-Tools incompatibilities and the most reliable solution I've found has been to compile glslang directly into the core VSG,  This has worked well and completely avoids the need of building VSG against 3rd party built glslang/SPIRV-Tools.

This is now merged with VSG master #729.  This means the VSG no longer required glslang as a dependency so will address all the previous issues highlighted in this thread.
